File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // void preorder(vector<int>& ans, TreeNode* root) {
    //     if(root == NULL) return;
    //     ans.push_back(root->val);
    //     preorder(ans, root->left);
    //     preorder(ans, root->right);
    // }

    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        // inorder(ans, root);
        // return ans;
        if(root == NULL) return ans;
        stack<TreeNode* >st;
        st.push(root);
        while(!st.empty()) {
            TreeNode* node = st.top();
            st.pop();
            if(node->right != NULL) st.push(node->right);
            if(node->left != NULL) st.push(node->left);
            ans.push_back(node->val);
        }
        return ans;
    }
};

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();

        if (root == null) {
            return ans;
        }

        Stack<TreeNode> st = new Stack<>();
        st.push(root);

        while (!st.isEmpty()) {
            TreeNode t = st.pop();
            ans.add(t.val);

            if (t.right != null) {
                st.push(t.right);
            }

            if (t.left != null) {
                st.push(t.left);
            }
        }

        return ans;
    }
}